network/linux_s/linux_s.c

```diff
@@ -0,0 +1,300 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <assert.h>
+#include "linux_s.h"
+#include "error.h"
+#include "util.h"
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <netinet/tcp.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include <string.h>
+#include <fcntl.h>
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+static int32_t linux_socket_open(struct linux_desc *desc, uint32_t *sock_id,
+				 enum socket_protocol prot, uint32_t buff_size);
+static int32_t linux_socket_close(struct linux_desc *desc, uint32_t sock_id);
+static int32_t linux_socket_connect(struct linux_desc *desc, uint32_t sock_id,
+				    struct socket_address *addr);
+static int32_t linux_socket_disconnect(struct linux_desc *desc,
+				       uint32_t sock_id);
+static int32_t linux_socket_send(struct linux_desc *desc, uint32_t sock_id,
+				 const void *data, uint32_t size);
+static int32_t linux_socket_recv(struct linux_desc *desc, uint32_t sock_id,
+				 void *data, uint32_t size);
+static int32_t linux_socket_sendto(struct linux_desc *desc, uint32_t sock_id,
+				   const void *data, uint32_t size,
+				   const struct socket_address* to);
+static int32_t linux_socket_recvfrom(struct linux_desc *desc, uint32_t sock_id,
+				     void *data, uint32_t size,
+				     struct socket_address *from);
+static int32_t linux_socket_bind(struct linux_desc *desc, uint32_t sock_id,
+				 uint16_t port);
+static int32_t linux_socket_listen(struct linux_desc *desc, uint32_t sock_id,
+				   uint32_t back_log);
+static int32_t linux_socket_accept(struct linux_desc *desc, uint32_t sock_id,
+				   uint32_t *client_socket_id);
+
+
+
+struct network_interface linux_net = {
```

Comment:
If you move the structure initialization at the end of the file you won't need to declare all functions.